### PR TITLE
Add EDMF driver for stable_bl_model.

### PR DESCRIPTION
**Pull request by ilopezgp**
### Description

This PR adds an EDMF driver for the simulation of the stable boundary layer. This is a dry simulation with zero/negligible area fraction. The PR also includes improvements of the EDMF helper functions to allow for simulations using a `DryModel`.

Co-authored by @charleskawczynski.

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

File: test/Atmos/EDMF/stable_bl_edmf.jl [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    N_up = n_updrafts(turbconv)
+    # GCM setting - Initialize the grid mean profiles of prognostic variables (ρ,e_int,q_tot,u,v,w)
+    z = altitude(m, aux)
+
+    # SCM setting - need to have separate cases coded and called from a folder - see what LES does
+    # a moist_thermo state is used here to convert the input θ,q_tot to e_int, q_tot profile
+    e_int = internal_energy(m, state, aux)
+
+    q_tot = state.moisture.ρq_tot / state.ρ
+    ts = PhaseEquil(m.param_set, e_int, state.ρ, q_tot)
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/stable_bl_edmf.jl:50]:
```suggestion
    ts = PhaseDry(m.param_set, e_int, state.ρ)
```

**Review by charleskawczynski — COMMENTED** [review]
We'll need to pass the MoistureModel from the driver if we want to run this in a purely dry mode.

File: experiments/AtmosLES/stable_bl_model.jl [hunk 2]
```diff
@@ -178,8 +178,11 @@ function init_problem!(problem, bl, state, aux, localgeo, t)
     π_exner = FT(1) - _grav / (c_p * θ) * z # exner pressure
     ρ = p0 / (R_gas * θ) * (π_exner)^(c_v / R_gas) # density
     # Establish thermodynamic state and moist phase partitioning
-    TS = PhaseEquil_ρθq(bl.param_set, ρ, θ_liq, q_tot)
-
+    if bl.moisture isa DryModel
+        TS = PhaseDry_ρθ(bl.param_set, ρ, T)
```
Comment (charleskawczynski) [inline experiments/AtmosLES/stable_bl_model.jl:182]:
```suggestion
        TS = PhaseDry_ρθ(bl.param_set, ρ, θ_liq)
```

**Comment by ilopezgp** [comment]
@charleskawczynski My last commit allows `stable_bl_les.jl` to be run with a DryModel() configuration of the `stable_bl_model.jl`. I had to remove the callback that used` q_tot`.

File: experiments/AtmosLES/stable_bl_model.jl [hunk 3]
```diff
@@ -202,9 +208,8 @@ end
 function surface_temperature_variation(state, t)
     FT = eltype(state)
     ρ = state.ρ
-    q_tot = state.moisture.ρq_tot / ρ
     θ_liq_sfc = FT(265) - FT(1 / 4) * (t / 3600)
-    TS = PhaseEquil_ρθq(param_set, ρ, θ_liq_sfc, q_tot)
+    TS = PhaseDry_ρθ(param_set, ρ, θ_liq_sfc)
```
Comment (yairchn) [inline experiments/AtmosLES/stable_bl_model.jl:212]: is there a need here for:

if bl.moisture isa DryModel
      TS = PhaseEquil_ρθq(param_set, ρ, θ_liq_sfc, q_tot)
else
      TS = PhaseDry_ρθ(param_set, ρ, θ_liq_sfc)
end

File: test/Atmos/EDMF/stable_bl_edmf.jl [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        varsindex(vsa, :turbconv),
+    )
+    # ---
+
+    dgn_config = config_diagnostics(driver_config)
+
+    cbtmarfilter = GenericCallbacks.EveryXSimulationSteps(1) do
+        Filters.apply!(
+            solver_config.Q,
+            ("moisture.ρq_tot", turbconv_filters(turbconv)...),
```
Comment (yairchn) [inline test/Atmos/EDMF/stable_bl_edmf.jl:172]: I think we can remove 
"moisture.ρq_tot"

File: experiments/AtmosLES/stable_bl_model.jl (hunk 3 above)
Comment (ilopezgp) [reply experiments/AtmosLES/stable_bl_model.jl:212]: Good point. This is something I also thought about. The problem is that `bl` is out of the scope of this function. Do you think it is worth modifying the inputs to the function to add the BalanceLaw @charleskawczynski?

File: test/Atmos/EDMF/stable_bl_edmf.jl [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+        up[i].ρaq_tot = gm.moisture.ρq_tot * a_min
+    end
+
+    # initialize environment covariance with zero for now
+    if z <= FT(250)
+        en.ρatke = gm.ρ * FT(0.4)*FT(1-z/250.0)*FT(1-z/250.0)*FT(1-z/250.0)
+    else
+        en.ρatke = FT(0)
+    end
+    en.ρaθ_liq_cv = FT(1e-5) / max(z, FT(10))
```
[the remark this answers is not in the document]
Comment (ilopezgp) [reply test/Atmos/EDMF/stable_bl_edmf.jl]: I changed it to the profiles we use in SCAMPy.

File: experiments/AtmosLES/stable_bl_model.jl (hunk 3 above)
Comment (charleskawczynski) [reply experiments/AtmosLES/stable_bl_model.jl:212]: I think it’s fine to add the balance law here. Should be changes in just this file.

File: experiments/AtmosLES/stable_bl_model.jl [hunk 6]
```diff
@@ -192,8 +197,9 @@ function init_problem!(problem, bl, state, aux, localgeo, t)
...
     state.ρu = SVector(ρu, ρv, ρw)
     state.ρe = ρe_tot
-    state.moisture.ρq_tot = ρ * q_tot
-
+    if !(bl.moisture isa DryModel)
+        state.moisture.ρq_tot = ρ * q_tot
+    end
     if z <= FT(50) # Add random perturbations to bottom 50m of model
         state.ρe += rand() * ρe_tot / 100
     end
```
Comment (yairchn) [inline experiments/AtmosLES/stable_bl_model.jl:205]: you are missing a call or initialize the EDMF here 
"init_state_prognostic!(bl.turbconv, bl, state, aux, localgeo, t)"

File: experiments/AtmosLES/stable_bl_model.jl [hunk 7]
```diff
@@ -192,19 +197,24 @@ function init_problem!(problem, bl, state, aux, localgeo, t)
...
+function surface_temperature_variation(state, t, moisture_model)
     FT = eltype(state)
     ρ = state.ρ
-    q_tot = state.moisture.ρq_tot / ρ
     θ_liq_sfc = FT(265) - FT(1 / 4) * (t / 3600)
-    TS = PhaseEquil_ρθq(param_set, ρ, θ_liq_sfc, q_tot)
+    if moisture_model == "dry"
+        TS = PhaseDry_ρθ(param_set, ρ, θ_liq_sfc)
+    else
+        TS = PhaseEquil_ρθq(param_set, ρ, θ_liq_sfc, q_tot)
```
Comment (yairchn) [inline experiments/AtmosLES/stable_bl_model.jl:216]: I think q_tot is not defined here

Comment (ilopezgp) [reply experiments/AtmosLES/stable_bl_model.jl:216]: Good catch

File: experiments/AtmosLES/stable_bl_model.jl [hunk 8]
```diff
@@ -252,15 +264,33 @@ function stable_bl_model(
...
     # Set up problem initial and boundary conditions
     if surface_flux == "prescribed"
         energy_bc = PrescribedEnergyFlux((state, aux, t) -> LHF + SHF)
         moisture_bc = PrescribedMoistureFlux((state, aux, t) -> moisture_flux)
     elseif surface_flux == "bulk"
         energy_bc = BulkFormulaEnergy(
             (state, aux, t, normPu_int) -> C_drag,
-            (state, aux, t) -> (surface_temperature_variation(state, t), q_sfc),
+            (state, aux, t) -> (
+                surface_temperature_variation(state, t, moisture_model),
```
Comment (charleskawczynski) [inline experiments/AtmosLES/stable_bl_model.jl:291]:
```suggestion
                surface_temperature_variation(state, t, moisture),
```

File: experiments/AtmosLES/stable_bl_model.jl [hunk 9]
```diff
@@ -192,19 +197,25 @@ function init_problem!(problem, bl, state, aux, localgeo, t)
...
+        state.moisture.ρq_tot = ρ * q_tot
+    end
     if z <= FT(50) # Add random perturbations to bottom 50m of model
         state.ρe += rand() * ρe_tot / 100
     end
+    init_state_prognostic!(bl.turbconv, bl, state, aux, localgeo, t)
 end
 
-function surface_temperature_variation(state, t)
+function surface_temperature_variation(state, t, moisture_model)
```
Comment (charleskawczynski) [inline experiments/AtmosLES/stable_bl_model.jl:209]:
```suggestion
function surface_temperature_variation(state, t, moisture)
```

File: experiments/AtmosLES/stable_bl_model.jl [hunk 10]
```diff
@@ -192,19 +197,25 @@ function init_problem!(problem, bl, state, aux, localgeo, t)
...
     FT = eltype(state)
     ρ = state.ρ
-    q_tot = state.moisture.ρq_tot / ρ
     θ_liq_sfc = FT(265) - FT(1 / 4) * (t / 3600)
-    TS = PhaseEquil_ρθq(param_set, ρ, θ_liq_sfc, q_tot)
+    if moisture_model == "dry"
```
Comment (charleskawczynski) [inline experiments/AtmosLES/stable_bl_model.jl:213]:
```suggestion
    if moisture isa DryModel
```

**Review by charleskawczynski — APPROVED** [review]
I think this looks fine. Should we remove the added `recover_thermo_state` methods? They never get called. Also, how do the results look for the new test?

**Comment by ilopezgp** [comment]
> I think this looks fine. Should we remove the added `recover_thermo_state` methods? They never get called. Also, how do the results look for the new test?

I will delete the `recover_thermo_state` methods. In terms of the results, they look physically reasonable for the first hour of simulation (which I have tested so far).

File: experiments/AtmosLES/stable_bl_model.jl [hunk 11]
```diff
@@ -252,15 +264,31 @@ function stable_bl_model(
...
     # Set up problem initial and boundary conditions
     if surface_flux == "prescribed"
         energy_bc = PrescribedEnergyFlux((state, aux, t) -> LHF + SHF)
         moisture_bc = PrescribedMoistureFlux((state, aux, t) -> moisture_flux)
     elseif surface_flux == "bulk"
         energy_bc = BulkFormulaEnergy(
             (state, aux, t, normPu_int) -> C_drag,
-            (state, aux, t) -> (surface_temperature_variation(state, t), q_sfc),
+            (state, aux, t) ->
+                (surface_temperature_variation(state, t, moisture), q_sfc),
```
Comment (charleskawczynski) [inline experiments/AtmosLES/stable_bl_model.jl:291]: Just realized that we’re doing a closure over `moisture`,  which could be causing the GPU failure.

File: test/Atmos/EDMF/stable_bl_edmf.jl [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+ClimateMachine.init(;
+    parse_clargs = true,
+    output_dir = get(ENV, "CLIMATEMACHINE_SETTINGS_OUTPUT_DIR", "output"),
+    fix_rng_seed = true,
+)
+using ClimateMachine.SingleStackUtils
+using ClimateMachine.Checkpoint
+using ClimateMachine.BalanceLaws: vars_state
+const clima_dir = dirname(dirname(pathof(ClimateMachine)));
+include(joinpath(clima_dir, "docs", "plothelpers.jl"));
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/stable_bl_edmf.jl:11]: Need to remove plot helpers here

File: experiments/AtmosLES/stable_bl_model.jl (hunk 11 above)
Comment (ilopezgp) [reply experiments/AtmosLES/stable_bl_model.jl:291]: how should I fix this?

**Pull request merged**